test/unit/sqs-redriver-test.js

```diff
@@ -0,0 +1,243 @@
+              delete_message_stub.should.not.be.called();
+            });
+          });
+
+        });
+
+        describe('when not possible to add messages to target queue', () => {
+          let mock_err;
+
+          beforeEach('mockReceiveOneMessage', () => {
+            first_message = mockIncompleteButSufficientSqsMessage('first');
+            receive_message_stub.onCall(0).yieldsAsync(null, [first_message]);
+            receive_message_stub.onCall(1).yieldsAsync(null, []);
+          });
+
+          beforeEach('mockSendMessage', () => {
+            const irrelevant_data = {};
+            send_message_stub.yieldsAsync(null, irrelevant_data);
+          });
+
+          beforeEach('mockFailedDelete', () => {
+            mock_err = new Error('Mocked AWS error');
+            delete_message_stub.yieldsAsync(mock_err, null);
+          });
+
+          it('should fail with AWS error', () => {
+            return redriver.redriveMessages(params).should.be.rejectedWith(mock_err);
+          });
+
+        });
+
+        describe('when message deletion fails', () => {
+          let mock_err;
+
+          beforeEach('mockReceiveOneMessage', () => {
+            first_message = mockIncompleteButSufficientSqsMessage('first');
+            receive_message_stub.onCall(0).yieldsAsync(null, [first_message]);
+            receive_message_stub.onCall(1).yieldsAsync(null, []);
+          });
+
+          beforeEach('mockFailedSend', () => {
+            mock_err = new Error('Mocked AWS error');
+            send_message_stub.yieldsAsync(mock_err, null);
+          });
+
+          it('should fail with AWS error', () => {
+            return redriver.redriveMessages(params).should.be.rejectedWith(mock_err);
+          });
+
+          it('should not delete any message from source queue', () => {
+            return redriver.redriveMessages(params).catch(() => {
+              delete_message_stub.should.not.be.called();
+            });
+          });
+
+        });
+
+      });
+
+    });
+
+    describe('with invalid params', () => {
+      let params;
+
+      beforeEach(() => {
+        params = createValidParams();
+      });
+
+      describe('missing sns_event', () => {
+
+        beforeEach(() => {
+          delete params.sns_event;
+        });
+
+        it('should fail with error', () => {
+          return redriver.redriveMessages(params).should.be.rejectedWith({
+            code: 422,
+          });
+        });
+
+      });
+
+      describe('with invalid message JSON', () => {
+
+        beforeEach(() => {
+          params.sns_event.Records[0].Sns.Message = {};
+        });
+
+        it('should fail with error', () => {
+          return redriver.redriveMessages(params).should.be.rejectedWith({
+            code: 422,
+          });
+        });
+
+      });
+
+    });
+
+  });
+
+});
+
+function createValidParams() {
+  const sns_event_copy = Object.assign({}, sns_event_fixture);
```

Comment:
This is a shallow copy. That's probably OK for this use case. Just wanted to point that out